### PR TITLE
Gpu support co2gaspvt

### PR DESCRIPTION
**Pull request by multitalentloes**
Adds test verifying that new classes work on the GPU, classes that are now tested include:
The test depends on new functionality in https://github.com/OPM/opm-common/pull/4222

- Constants
- IdealGas
- Brine_CO2
- UniformTabulated2DFunction
- BrineDynamic
- CO2
- SimpleHuDuanH20
- Co2GasPvt

### CODE REVIEW

File: opm/simulators/linalg/gpuistl/GpuBuffer.cpp [hunk 1]
```diff
@@ -33,6 +33,14 @@ GpuBuffer<T>::GpuBuffer(const std::vector<T>& data)
 {
 }
 
+// some classes need a default constructor...
+// should we reintroduce the default constructor that allocates 0 bytes and initializes the handles?
```
Comment (atgeirr) [inline opm/simulators/linalg/gpuistl/GpuBuffer.cpp:37]: There is already a declared default constructor, but it seems it is not actually implemented before this, which is weird.

That aside, I think it should not allocate anything. Default construct then resize should allocate only once. I suggest removing this implementation, defaulting `m_numberOfElements` to zero in the header file, and then just defaulting the constructor in the header:
```
    /**
     * @brief Default constructor that will initialize cublas and allocate 0 bytes of memory
     */
    GpuBuffer() = default;
```
That way you do not need to change the doc string either to say 1 instead 0...

Finally, note that it is currently declared explicit in the header, which is unnecessary and misleading.

File: CMakeLists_files.cmake [hunk 2]
```diff
@@ -457,12 +457,14 @@ if (HAVE_CUDA)
...
   ADD_CUDA_OR_HIP_FILE(TEST_SOURCE_FILES tests test_gpu_ad.cu)
   ADD_CUDA_OR_HIP_FILE(TEST_SOURCE_FILES tests test_gpu_linear_two_phase_material.cu)
+  ADD_CUDA_OR_HIP_FILE(TEST_SOURCE_FILES tests test_gpuPvt.cu)
 
   # for loop providing the flag --expt-relaxed-constexpr to fix some cuda issues with constexpr
   if(NOT CONVERT_CUDA_TO_HIP)
     set(CU_FILES_NEEDING_RELAXED_CONSTEXPR
       tests/gpuistl/test_gpu_ad.cu
       tests/gpuistl/test_gpu_linear_two_phase_material.cu
+      tests/gpuistl/test_co2gaspvt.cu
```
Comment (atgeirr) [inline CMakeLists_files.cmake:467]: Did you mean test_gpuPvt.cu?

File: tests/gpuistl/test_gpuPvt.cu [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+
+using Evaluation = Opm::DenseAd::Evaluation<float, 3>;
+using GpuB = const Opm::gpuistl::GpuBuffer<double>;
+using GpuV = Opm::gpuistl::GpuView<const double>;
+
+using GpuTab = Opm::UniformTabulated2DFunction<double, GpuV>;
+
+using GpuBufCo2Tables = Opm::CO2Tables<double, GpuB>;
+using GpuViewCO2Tables = Opm::CO2Tables<double, GpuV>;
+template Opm::CO2Tables<double, GpuB>::CO2Tables(const Opm::UniformTabulated2DFunction<double, GpuB>&, const Opm::UniformTabulated2DFunction<double, GpuB>&);
```
Comment (atgeirr) [inline tests/gpuistl/test_gpuPvt.cu:35]: What do these explicit instantiations do? We do not access this in a scope linking to this file.

File: tests/gpuistl/test_gpuPvt.cu [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+#include <opm/simulators/linalg/gpuistl/detail/gpu_safe_call.hpp>
+#include <opm/simulators/linalg/gpuistl/GpuBuffer.hpp>
+#include <opm/simulators/linalg/gpuistl/GpuView.hpp>
+
+#include <cuda_runtime.h>
+#include <vector>
+#include <utility>
+#include <cmath>
+
+using Evaluation = Opm::DenseAd::Evaluation<float, 3>;
```
Comment (atgeirr) [inline tests/gpuistl/test_gpuPvt.cu:27]: Is float used on purpose here? Everything else is doubles.

File: tests/gpuistl/test_gpuPvt.cu [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+
+    // Launch the kernel
+    kernel<<<1, 1>>>(args..., resultOnGpu);
+
+    // Check for any errors in kernel launch
+    OPM_GPU_SAFE_CALL(cudaPeekAtLastError());
+    OPM_GPU_SAFE_CALL(cudaDeviceSynchronize());
+
+    // Retrieve the result from the GPU to the CPU
+    OPM_GPU_SAFE_CALL(cudaMemcpy(&gpuComputedResultOnCpu, resultOnGpu, sizeof(double), cudaMemcpyDeviceToHost));
```
Comment (atgeirr) [inline tests/gpuistl/test_gpuPvt.cu:132]: What would `gpuComputedResultOnCpu = *resultOnGpu;` do?

**Review by atgeirr — COMMENTED** [review]
Only a few minor issues to be resolved. Should be merged when fixed.

File: tests/gpuistl/test_gpuPvt.cu (hunk 5 above)
Comment (multitalentloes) [reply tests/gpuistl/test_gpuPvt.cu:132]: That would not yield anything meaningful, you cannot dereference a gpu pointer and get the value in CPU code

File: CMakeLists_files.cmake (hunk 2 above)
Comment (multitalentloes) [reply CMakeLists_files.cmake:467]: Good catch, I decided to change the name to `test_gpuPvt.cu` at some point

**Comment by multitalentloes** [comment]
jenkins build this please

File: opm/simulators/linalg/gpuistl/GpuBuffer.cpp (hunk 1 above)
Comment (multitalentloes) [reply opm/simulators/linalg/gpuistl/GpuBuffer.cpp]: fixed now with updated doc string and logic in resize function accounting for original size being 0

**Review by atgeirr — APPROVED** [review]

**Pull request merged**